dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ProgramEncoder.java

```diff
@@ -258,29 +255,113 @@ private BooleanFormula encodeConsistentThreadCF(Thread thread) {
+            for (ControlBarrier barrier : entry.getValue()) {
+                enc = bmgr.and(enc, bmgr.equivalence(variable, context.execution(barrier)));
+            }
+        }
+        return enc;
+    }
+
+    private BooleanFormula encodeNamedControlBarriers() {
+        BooleanFormulaManager bmgr = context.getBooleanFormulaManager();
+        BooleanFormula enc = bmgr.makeTrue();
+
+        Map<String, List<NamedBarrier>> barriers = context.getTask().getProgram().getThreadEvents(NamedBarrier.class).stream()
+                .collect(groupingBy(b -> getWorkgroupId(b.getThread()) + "@" + b.getInstanceId()));
+
+        for (List<NamedBarrier> events : barriers.values()) {
+            for (NamedBarrier e1 : events) {
+                List<NamedBarrier> other = events;
+                if (e1.getResourceId() instanceof IntLiteral) {
+                    other = events.stream()
+                            .filter(e2 -> !(e2.getResourceId() instanceof IntLiteral) || e1.getResourceId().equals(e2.getResourceId()))
+                            .toList();
+                }
+                if (e1.getQuorum() != null) {
+                    enc = bmgr.and(enc, encodeNamedBarrierCfQuorum(e1, other));
+                } else {
+                    enc = bmgr.and(enc, encodeNamedBarrierCfAll(e1, other));
                 }
-                enc = bmgr.and(enc, bmgr.equivalence(allCF, context.execution(e1)));
             }
         }
         return enc;
     }
 
+    private BooleanFormula encodeNamedBarrierCfAll(NamedBarrier e1, List<NamedBarrier> events) {
+        BooleanFormulaManager bmgr = context.getBooleanFormulaManager();
+        Expression id1 = e1.getResourceId();
+        BooleanFormula allCF = bmgr.makeTrue();
+        for (NamedBarrier e2 : events) {
+            BooleanFormula sameId = context.equal(context.encodeExpressionAt(id1, e1), context.encodeExpressionAt(e2.getResourceId(), e2));
+            BooleanFormula cf = bmgr.or(context.controlFlow(e2), bmgr.not(sameId));
+            allCF = bmgr.and(allCF, cf);
+        }
+        return bmgr.and(context.sync(e1), bmgr.equivalence(allCF, context.execution(e1)));
+    }
+
+    /*
+     * Variables:
+     * - `bool: sync(e: NamedBarrier)` : defines whether `e` is in `syncbar` relation
+     * - `int: quorum(e: NamedBarrier)` : a parameter of `e` defining the minimal number of threads to reach the barrier
+     * - `cf_count(e: NamedBarrier)` : the number of threads that reached the barrier in control flow
+     * - `sync_count(e: NamedBarrier)` : the number of threads that synchronize on the barrier
+     *
+     * Note: `cf_count` and `sync_count` cannot use a single variable for the whole barrier and must be defined for each
+     * event, because for PTX `resourceId` may vary in different executions.
+     *
+     * Constraints:
+     * - `sync(e) => (cf(e) /\ exec(e))`
+     * - `exec(e) <=> (cf(e) /\ (cf_count(e) >= quorum(e)))`
```

Comment:
If I am understanding the code correctly it should be
```suggestion
     * - `exec(e) <=> (cf(e) /\ (syncCount(e) >= quorum(e)))`
```

Reply:
Fixed, thanks